jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1297 @@
+        {
+            throw new UnsupportedOperationException(UNMODIFIABLE);
+        }
+
+        /**
+         * Creates a <Code>Spliterator</code> over the elements in the set's unmodifiable copy.
+         *
+         * @return a  <code>Spliterator</code> over the elements in the backing set's unmodifiable copy.
+         */
+        @Override
+        public Spliterator<E> spliterator()
+        {
+            return getCopy().spliterator();
+        }
+
+        /**
+         * Return a sequential <code>Stream</code> with the backing set's unmodifiable copy as its
+         * source.
+         * @return a sequential <code>Stream</code> with the backing set's unmodifiable copy as its
+         * source.
+         */
+        @Override
+        public Stream<E> stream()
+        {
+            return getCopy().stream();
+        }
+
+        /**
+         * Return a possibly parallel <code>Stream</code> with the backing set's unmodifiable copy as
+         * its source.
+         * @return a possibly parallel <code>Stream</code> with the backing set's unmodifiable copy
+         * as its source.
+         */
+        @Override
+        public Stream<E> parallelStream()
+        {
+            return getCopy().parallelStream();
+        }
+
+        /**
+         * Compares the specified object with this set for equality.
+         * @param o object to be compared for equality with this set.
+         * @return <code>true</code> if o and this set are the same object or o is equal to the
+         * backing object, false otherwise.
+         */
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o)
+                return true;
+            readWriteLock.readLock().lock();
+            try {
+                return set.equals(o);
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
+        }
+
+        /**
+         * Return the backing set's hashcode.
+         * @return the backing set's hashcode.
+         */
+        @Override
+        public int hashCode()
+        {
+            readWriteLock.readLock().lock();
+            try {
+                return set.hashCode();
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
+        }
+
+        /**
+         * Return the backing set's toString result.
+         * @return the backing set's toString result.
+         */
+        @Override
+        public String toString()
+        {
+            readWriteLock.readLock().lock();
+            try {
+                return set.toString();
+            } finally {
+                readWriteLock.readLock().unlock();
+            }
+        }
+
+        /**
+         * Get the backing set's unmodifiable copy.
+         * @return the backing set's unmodifiable copy.
+         */
+        private Set<E> getCopy()
+        {
+            readWriteLock.readLock().lock();
+            try {
+                if (copy != null)
```

Comment:
You're using the "double-checked locking" idiom here, which is OK since Java 1.5's memory model, but you haven't implemented it correctly.

See http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

And the optimized version in https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

Reply:
Wow, thanks for the tip.